### DISCUSSION TITLE
SSR and components

### DISCUSSION THREAD

**Discussion by hkirsman** [vercel/next.js#19113] [Help] [answered]
Hello!

If anybody could explain.. Can and should I be able to create component that fetches data when the page is rendered in server so that the markup is there already during page load? I'm trying to build menu component but the data is fetched client side. Is it meant to be like that components by default are fetching data client side? Is it a hack to try to make it render on server side?

Thank you!

**Comment by jamesmosier** [comment, marked as answer]
Components that fetch data (via xhr or fetch) only run on the client, not the server. 

Next has a number of methods you could use to fetch data on the server though: https://nextjs.org/docs/basic-features/data-fetching

* `getStaticProps` can be used to make requests during your build and then those props can be passed to your page
* `getServerSideProps` is similar but this runs on every request to your page from a user. As the name infers this runs on the server.

**Comment by hkirsman** [comment]
But components by default are meant to work on client side?

**Reply by jamesmosier** [reply]
Component can work on both the client and server side the same.

When you are fetching data via xhr (`axios` for example) or `fetch` then that will only happen on the client side and **not** the server.

**Comment by hkirsman** [comment]
I think the way I have project set up it should work both ways. Maybe I should add some _if_ or hook to my component to make sure data is pulled on server.
